Public/Src/Cache/ContentStore/Distributed/NuCache/ContentLocationDatabase.cs

```diff
@@ -355,19 +456,96 @@ public BoolResult RestoreCheckpoint(OperationContext context, AbsolutePath check
+        protected abstract bool TryGetEntryCoreFromStorage(OperationContext context, ShortHash hash, out ContentLocationEntry entry);
 
         /// <nodoc />
-        protected abstract void Store(OperationContext context, ShortHash hash, ContentLocationEntry entry);
+        protected bool TryGetEntryCore(OperationContext context, ShortHash hash, out ContentLocationEntry entry)
+        {
+            Counters[ContentLocationDatabaseCounters.NumberOfGetOperations].Increment();
+
+            if (IsInMemoryCacheEnabled && _inMemoryCache.TryGetEntry(hash, out entry))
+            {
+                return true;
+            }
+
+            return TryGetEntryCoreFromStorage(context, hash, out entry);
+        }
+
+        /// <nodoc />
+        internal abstract void Persist(OperationContext context, ShortHash hash, ContentLocationEntry entry);
 
         /// <nodoc />
-        protected abstract void Delete(OperationContext context, ShortHash hash);
+        internal virtual void PersistBatch(OperationContext context, IEnumerable<KeyValuePair<ShortHash, ContentLocationEntry>> pairs)
+        {
+            foreach (var pair in pairs)
+            {
+                Persist(context, pair.Key, pair.Value);
+            }
+        }
+
+        /// <nodoc />
+        protected void Store(OperationContext context, ShortHash hash, ContentLocationEntry entry)
+        {
+            Counters[ContentLocationDatabaseCounters.NumberOfStoreOperations].Increment();
+
+            if (IsInMemoryCacheEnabled)
+            {
+                _inMemoryCache.Store(context, hash, entry);
+
+                // The fact that this is == is important to ensure it can only be triggered once by this condition
+                if (Interlocked.Increment(ref _cacheUpdatesSinceLastFlush) == _configuration.CacheMaximumUpdatesPerFlush)
+                {
+                    Counters[ContentLocationDatabaseCounters.NumberOfCacheFlushesTriggeredByUpdates].Increment();
+                    ForceCacheFlushAsync(context).FireAndForget(context);
+                }
+            }
+            else
+            {
+                Persist(context, hash, entry);
+                Counters[ContentLocationDatabaseCounters.NumberOfPersistedEntries].Increment();
+            }
+        }
+
+        /// <nodoc />
+        protected void Delete(OperationContext context, ShortHash hash)
+        {
+            Store(context, hash, entry: null);
+        }
+
+        private Task ForceCacheFlushAsync(OperationContext context)
+        {
+            if (!IsInMemoryCacheEnabled)
+            {
+                return BoolResult.SuccessTask;
+            }
+
+            return context.PerformOperationAsync(
+                Tracer,
+                async () =>
+                {
+                    try
+                    {
+                        await _inMemoryCache.FlushAsync(context);

```

Comment:
>FlushAsync [](start = 45, length = 10)

Its worth noting that this could execute synchronously. Maybe we should do an await Task.Yield() above.